### PR TITLE
Don't ignore values if default on export

### PR DESCRIPTION
**Pull request by timja**
GlobalConfiguration descriptors call load() in them which loads the XML
file, this causes
https://github.com/jenkinsci/configuration-as-code-plugin/blob/46dd452244128b0f40496d4483b2cc5f66a8de0e/plugin/src/main/java/io/jenkins/plugins/casc/impl/configurators/DescriptorConfigurator.java#L60

to create the reference instance which is the same as the actual
instance and therefore prevents these GlobalConfiguration's from being exported

Fixes: https://github.com/jenkinsci/configuration-as-code-plugin/issues/729

I've tested this on azure-keyvault-plugin and timestamper-plugin

Tests pass there now with this change

Thoughts @ewelinawilkosz / @casz  / @ndeloof 

Example default file after this change:
```
configuration-as-code:
  deprecated: reject
  restricted: reject
  unknown: reject
  version: ONE
jenkins:
  agentProtocols:
  - "CLI-connect"
  - "CLI2-connect"
  - "JNLP-connect"
  - "JNLP2-connect"
  - "JNLP4-connect"
  - "Ping"
  crumbIssuer: "FAILED TO EXPORT hudson.model.Hudson#crumbIssuer: \nio.jenkins.plugins.casc.ConfiguratorException:\
    \ Cannot find configurator for type class org.jvnet.hudson.test.TestCrumbIssuer\n\
    \tat io.jenkins.plugins.casc.impl.DefaultConfiguratorRegistry$1.load(DefaultConfiguratorRegistry.java:97)\n\
    \tat io.jenkins.plugins.casc.impl.DefaultConfiguratorRegistry$1.load(DefaultConfiguratorRegistry.java:93)\n\
    \tat com.google.common.cache.LocalCache$LoadingValueReference.loadFuture(LocalCache.java:3568)\n\
    \tat com.google.common.cache.LocalCache$Segment.loadSync(LocalCache.java:2350)\n\
    \tat com.google.common.cache.LocalCache$Segment.lockedGetOrLoad(LocalCache.java:2313)\n\
    \tat com.google.common.cache.LocalCache$Segment.get(LocalCache.java:2228)\n\t\
    at com.google.common.cache.LocalCache.get(LocalCache.java:3965)\n\tat com.google.common.cache.LocalCache.getOrLoad(LocalCache.java:3969)\n\
    \tat com.google.common.cache.LocalCache$LocalManualCache.get(LocalCache.java:4829)\n\
    \tat io.jenkins.plugins.casc.impl.DefaultConfiguratorRegistry.lookupOrFail(DefaultConfiguratorRegistry.java:70)\n\
    \tat io.jenkins.plugins.casc.ConfigurationContext.lookupOrFail(ConfigurationContext.java:71)\n\
    \tat io.jenkins.plugins.casc.impl.configurators.HeteroDescribableConfigurator.describe(HeteroDescribableConfigurator.java:164)\n\
    \tat io.jenkins.plugins.casc.impl.configurators.HeteroDescribableConfigurator.describe(HeteroDescribableConfigurator.java:44)\n\
    \tat io.jenkins.plugins.casc.Attribute.describe(Attribute.java:196)\n\tat io.jenkins.plugins.casc.Configurator.describe(Configurator.java:163)\n\
    \tat io.jenkins.plugins.casc.ConfigurationAsCode.export(ConfigurationAsCode.java:405)\n\
    \tat hudson.plugins.timestamper.ConfigAsCodeTest.export_configuration(ConfigAsCodeTest.java:37)\n\
    \tat sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)\n\tat sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)\n\
    \tat sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)\n\
    \tat java.lang.reflect.Method.invoke(Method.java:498)\n\tat org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)\n\
    \tat org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)\n\
    \tat org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)\n\
    \tat org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)\n\
    \tat org.jvnet.hudson.test.JenkinsRule$1.evaluate(JenkinsRule.java:552)\n\tat\
    \ org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:298)\n\
    \tat org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:292)\n\
    \tat java.util.concurrent.FutureTask.run(FutureTask.java:266)\n\tat java.lang.Thread.run(Thread.java:748)\n"
  disableRememberMe: false
  mode: NORMAL
  numExecutors: 2
  primaryView:
    all:
      name: "all"
      properties: "FAILED TO EXPORT hudson.model.AllView#properties: No configurator\
        \ found for type class hudson.model.ViewProperty"
  quietPeriod: 5
  remotingSecurity:
    enabled: false
  scmCheckoutRetryCount: 0
  slaveAgentPort: 0
  updateCenter:
    sites:
    - id: "default"
      url: "http://localhost:63836/update-center.json"
  views:
  - all:
      name: "all"
      properties: "FAILED TO EXPORT hudson.model.AllView#properties: No configurator\
        \ found for type class hudson.model.ViewProperty"
security:
  downloadSettings:
    useBrowser: false
  remotingCLI:
    enabled: true
  sSHD:
    port: -1
unclassified:
  location:
    adminAddress: "address not configured yet <nobody@nowhere>"
    url: "http://localhost:63818/jenkins/"
  pollSCM:
    pollingThreadCount: 10
  timestamperConfig:
    allPipelines: true
    elapsedTimeFormat: "'<b>'HH:mm:ss'</b> '"
    systemTimeFormat: "'<b>'mm:ss'</b> '"
tool:
  jdk:
    installations:
    - home: "/Library/Java/JavaVirtualMachines/jdk1.8.0_152.jdk/Contents/Home/jre"
      name: "default"
```

### CODE REVIEW

**Comment by casz** [comment]
Since it is a PR that would make it a fix 😅 Wonder what happen if you label it with both bug and fix... Making it a bugfix :sweat_smile:

**Comment by timja** [comment]
Jenkins is working but not Codacy now 😉 
image

**Comment by casz** [comment]
ARGH, why must you face a white GitHub upon me :scream:
Save your eyes with https://github.com/StylishThemes/GitHub-Dark

**Comment by casz** [comment]
I believe Codacy has an ongoing maintenance window.
https://status.codacy.com/maintenances/krP0XcSMNBIW

**Comment by timja** [comment]
the maintenance window hasn't started yet, first thing tomorrow morning the check is there now but not loading when I click it.

I have the dark theme installed but it was it bit intense for me,
and I didn't like all the avatars had white borders around them 

I have dark mode on my mac and chrome 👅

**Comment by casz** [comment]
@timja perhaps the actual fix is to do just like in JCasC, have an empty constructor that calls load and one that uses databound constructor :smile: 

https://github.com/jenkinsci/configuration-as-code-plugin/blob/63c82a8b64bb5b1f5053276970b27b577a2f0150/plugin/src/main/java/io/jenkins/plugins/casc/CasCGlobalConfig.java#L19-L26

**Comment by timja** [comment]
That assumes that you have a mandatory field though...
doesn't fix it for those that don't

**Comment by timja** [comment]
i.e. azure-keyvault-plugin,

You can set a global url, or a credential, or both.

None are required though

**Comment by casz** [comment]
Seems like compare become useless and is used to describe perhaps best to remove the method and use the default describe method :laughing:

**Comment by timja** [comment]
It's useless but it is part of the public API

unfortunately the methods return different types

One a CNode and one a Mapping

Can have a look at removing and see the impact

**Comment by casz** [comment]
@timja perhaps the best thing is to remove the compare call inside the describe block of `DescriptorConfiguratior`. This avoids messing with the whole chain of other configurators that uses compare and reference object.

**Comment by casz** [comment]
Feel free to revert and force push my changes away :smile:

**Comment by timja** [comment]
squash will sort that out

File: plugin/src/main/java/io/jenkins/plugins/casc/BaseConfigurator.java [hunk 1]
```diff
@@ -380,7 +380,13 @@ protected final void handleUnknown(Mapping config, ConfigurationContext context)
     }
 
     protected @NonNull Mapping compare(T instance, T reference, ConfigurationContext context) throws Exception {
-        return attributesToMapping(instance, context);
+
```
Comment (casz) [inline plugin/src/main/java/io/jenkins/plugins/casc/BaseConfigurator.java]: I don't care for the :ghost: whitspace

**Comment by casz** [comment]
Would have loved a downstream PR but Jenkins CI did not feel like cooperating with an incremental hpi file.

**Comment by timja** [comment]
I've got two PRs ready to go locally,
but meh, CI didn't want to give us a windows node so it was stuck for 12 hours...

**Comment by timja** [comment]
Will open the PRs later if incrementals pass on master

**Comment by darxriggs** [comment]
Are you aware of https://javadoc.jenkins-ci.org/hudson/model/PersistentDescriptor.html?

**Comment by casz** [comment]
What do you mean by that @darxriggs? :thinking: Yes, I am aware of it. that doesn't stop all the plugins out there from calling Load in their constructor :sweat_smile:

The issue was JCasC saw the object as the default. Since it had the same fields as what JCasC had in its YAML mapping hence it was excluded from export.

**Comment by darxriggs** [comment]
`PersistentDescriptor` can also be used to fix exporting for JCasC - see e.g. https://github.com/jenkinsci/ivy-plugin/commit/908bf53c716e33e710c4940fbef552bbf648dcf3

**Comment by timja** [comment]
2.140 is quite a high baseline, much easier to fix this in JCasC.

**Comment by darxriggs** [comment]
Yes, regarding 2.140 I am with you.

**Comment by darxriggs** [comment]
Are all values or just non-default ones exported now after this change?
That's not obvious for me looking a this change.

Personally I would like to have all values exported (or an option to decide).
Because I think it's much more easier for a user to adapt a fully exported YAML instead of having to look at the generated documentation to find the fields to add to a YAML that contains only the non-default values.

**Comment by timja** [comment]
Anything that was getting exported by `DescriptorConfigurator` is now exported, previously they were only exported it they weren't the default.

Agreed

**Pull request merged**